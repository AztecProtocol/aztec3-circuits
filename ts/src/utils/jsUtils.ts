export function range(n: number, offset: number = 0) {
  const ret: number[] = [];
  for (let i = 0; i < n; i++) {
    ret.push(offset + i);
  }
  return ret;
}

export function assertLength(obj: any, member: string, length: number) {
  if (obj[member].length !== length) {
    throw new Error(
      `Expected ${member} to have length ${length}! Was: ${obj[member].length}`
    );
  }
}
