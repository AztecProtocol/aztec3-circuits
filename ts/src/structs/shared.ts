import { randomBytes } from "crypto";
import { serializeToBuffer } from "../wasm/serialize.js";
import { checkLength } from "./utils.js";

export class Fr {
  static SIZE_IN_BYTES = 32;

  constructor(private buffer: Buffer) {
    if (buffer.length != Fr.SIZE_IN_BYTES) {
      throw new Error(
        `Unexpected buffer size ${buffer.length} (expected ${Fr.SIZE_IN_BYTES} bytes)`
      );
    }
  }

  static random() {
    return new Fr(
      Buffer.concat([Buffer.alloc(1, 0), randomBytes(Fr.SIZE_IN_BYTES - 1)])
    );
  }

  toString() {
    return "0x" + this.buffer.toString("hex");
  }

  toBuffer() {
    return this.buffer;
  }
}

/**
 * For Ethereum addresses, which must be treated as 32 bytes.
 * @param buffer - The 20 byte ethereum address buffer.
 * @returns The 32 byte padded buffer.
 */
function pad32(buffer: Buffer) {
  // Create a 32-byte Buffer filled with zeros
  const paddedBuffer = Buffer.alloc(32);

  // Calculate the padding length
  const paddingLength = paddedBuffer.length - buffer.length;

  // Copy the original Buffer into the padded Buffer with an offset
  buffer.copy(paddedBuffer, paddingLength);
  return paddedBuffer;
}

export class EthAddress {
  static SIZE_IN_BYTES = 20;

  constructor(private buffer: Buffer) {
    if (buffer.length != EthAddress.SIZE_IN_BYTES) {
      throw new Error(
        `Unexpected buffer size ${buffer.length} (expected ${EthAddress.SIZE_IN_BYTES} bytes)`
      );
    }
  }

  static random() {
    return new Fr(randomBytes(EthAddress.SIZE_IN_BYTES));
  }

  toString() {
    return "0x" + this.buffer.toString("hex");
  }

  toBuffer() {
    return pad32(this.buffer);
  }
}

export class MembershipWitness<N extends number> {
  constructor(pathSize: N, public leafIndex: UInt32, public siblingPath: Fr[]) {
    checkLength(this.siblingPath, pathSize, "MembershipWitness.siblingPath");
  }

  toBuffer() {
    return serializeToBuffer(this.leafIndex, ...this.siblingPath);
  }
}

export class AggregationObject {
  public hasData: false = false;
  constructor(
    public p0: G1,
    public p1: G1,
    public publicInputs: Fr[],
    public proofWitnessIndices: UInt32[]
  ) {}

  toBuffer() {
    return serializeToBuffer(
      this.p0,
      this.p1,
      this.publicInputs,
      this.proofWitnessIndices,
      this.hasData
    );
  }
}

export type UInt32 = number;

// TODO: Define proper type for AztecAddress
export type AztecAddress = Fr;

// TODO: What is a Curve::G1?
export type G1 = Fr;

export type Proof = Buffer;
export type VK = Buffer;
