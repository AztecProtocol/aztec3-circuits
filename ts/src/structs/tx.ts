import { serializeToBuffer } from "../wasm/serialize.js";
import { Fr } from "./shared.js";

/**
 * Contract deployment data in a @TxContext.
 * cpp/src/aztec3/circuits/abis/contract_deployment_data.hpp
 * 
 * @todo Used the Cpp type as source of truth, which does not match the specification.
 * Spec includes constructorVkHash but does not have contract_data_hash nor constructor_hash.
 */
export class ContractDeploymentData {
  constructor (
    public contractDataHash: Fr,
    public functionTreeRoot: Fr,
    public constructorHash: Fr,
    public contractAddressSalt: Fr,
    /**
     * EthAddress of the Portal Contract
     * TODO: Change to EthAddress once there's a uint160 circuit type
     */
    public portalContractAddress: Fr,
  ) { }

  toBuffer() {
    return serializeToBuffer(
      this.contractDataHash,
      this.functionTreeRoot,
      this.constructorHash,
      this.contractAddressSalt,
      this.portalContractAddress,
    );
  }
}

/**
 * Transaction context
 * cpp/src/aztec3/circuits/abis/tx_context.hpp
 */
export class TxContext {
  constructor (
    public isFeePaymentTx: false,
    public isRebatePaymentTx: false,
    public isContractDeployment: true,
    public contractDeploymentData: ContractDeploymentData,
    public referenceBlockNumber: Fr,
  ) {}

  toBuffer() {
    return serializeToBuffer(
      this.isFeePaymentTx,
      this.isRebatePaymentTx,
      this.isContractDeployment,
      this.contractDeploymentData,
      this.referenceBlockNumber,
    );
  }
}