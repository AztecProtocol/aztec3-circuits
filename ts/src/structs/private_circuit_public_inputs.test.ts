import { expectSerializeToMatchSnapshot } from "../tests/expectSerializeToMatchSnapshot.js";
import { asBEBuffer, fr } from "../tests/testUtils.js";
import { range } from "../utils/jsUtils.js";
import { CallContext } from "./call_context.js";
import {
  ARGS_LENGTH,
  EMITTED_EVENTS_LENGTH,
  RETURN_VALUES_LENGTH,
  NEW_COMMITMENTS_LENGTH,
  NEW_NULLIFIERS_LENGTH,
  PRIVATE_CALL_STACK_LENGTH,
  PUBLIC_CALL_STACK_LENGTH,
  L1_MSG_STACK_LENGTH,
} from "./constants.js";
import { PrivateCircuitPublicInputs } from "./private_circuit_public_inputs.js";
import { EthAddress, Fr } from "./shared.js";
import { ContractDeploymentData } from "./tx.js";

/**
 * Create sequential test data for ContractDeploymentData.
 * @returns Test data.
 */
function contractDeploymentData() {
  return new ContractDeploymentData(
    new Fr(asBEBuffer(1)),
    new Fr(asBEBuffer(2)),
    new Fr(asBEBuffer(3)),
    new Fr(asBEBuffer(4))
  );
}

/**
 * Create sequential test data for PrivateCircuitPublicInputs.
 * @returns Test data.
 */
function privateCircuitPublicInputs() {
  return PrivateCircuitPublicInputs.from({
    callContext: new CallContext(
      asBEBuffer(1),
      asBEBuffer(2),
      new EthAddress(asBEBuffer(3, /* eth address is 20 bytes */ 20)),
      true,
      true,
      true
    ),
    args: range(ARGS_LENGTH).map(fr),
    emittedEvents: range(EMITTED_EVENTS_LENGTH, 0x100).map(fr), // TODO not in spec
    returnValues: range(RETURN_VALUES_LENGTH, 0x200).map(fr),
    newCommitments: range(NEW_COMMITMENTS_LENGTH, 0x300).map(fr),
    newNullifiers: range(NEW_NULLIFIERS_LENGTH, 0x400).map(fr),
    privateCallStack: range(PRIVATE_CALL_STACK_LENGTH, 0x500).map(fr),
    publicCallStack: range(PUBLIC_CALL_STACK_LENGTH, 0x600).map(fr),
    l1MsgStack: range(L1_MSG_STACK_LENGTH, 0x700).map(fr),
    historicContractTreeRoot: new Fr(asBEBuffer(0x801)), // TODO not in spec
    historicPrivateDataTreeRoot: new Fr(asBEBuffer(0x901)),
    historicPrivateNullifierTreeRoot: new Fr(asBEBuffer(0x1001)), // TODO not in spec
    contractDeploymentData: contractDeploymentData(),
  });
}

describe("basic PrivateCircuitPublicInputs serialization", () => {
  it(`serializes a trivial PrivateCircuitPublicInputs and prints it`, async () => {
    // Test the data case: writing (mostly) sequential numbers
    await expectSerializeToMatchSnapshot(
      privateCircuitPublicInputs().toBuffer(),
      "abis__test_roundtrip_serialize_private_circuit_public_inputs"
    );
  });
});
