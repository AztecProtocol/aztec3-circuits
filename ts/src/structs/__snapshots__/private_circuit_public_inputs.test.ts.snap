// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic PrivateCircuitPublicInputs serialization serializes a trivial PrivateCircuitPublicInputs and prints it 1`] = `
"call_context: msg_sender: 0x1
storage_contract_address: 0x2
tx_origin: 0x3
is_delegate_call: 1
is_static_call: 1
is_contract_deployment: 1

args: [ 0x1 0x2 0x3 0x4 0x5 0x6 0x7 0x8 ]
return_values: [ 0x201 0x202 0x203 0x204 ]
emitted_events: [ 0x101 0x102 0x103 0x104 ]
new_commitments: [ 0x301 0x302 0x303 0x304 ]
new_nullifiers: [ 0x401 0x402 0x403 0x404 ]
private_call_stack: [ 0x501 0x502 0x503 0x504 ]
public_call_stack: [ 0x601 0x602 0x603 0x604 ]
l1_msg_stack: [ 0x701 0x702 ]
historic_private_data_tree_root: 0x901
historic_nullifier_tree_root: 0x1001
contract_deployment_data: constructor_vk_hash: 0x1
function_tree_root: 0x2
contract_address_salt: 0x3
portal_contract_address: 0x4

contract_deployment_data: constructor_vk_hash: 0x1
function_tree_root: 0x2
contract_address_salt: 0x3
portal_contract_address: 0x4

"
`;
